packages/x-date-pickers/src/internals/hooks/useField/useFieldV7TextField.ts

```diff
@@ -176,146 +145,60 @@ export const useFieldV7TextField = <
-
-        if (cursorPosition === 0) {
-          setSelectedSections(sectionOrder.startIndex);
-          return;
-        }
-
-        let sectionIndex = 0;
-        let cursorOnStartOfSection = 0;
-
-        while (cursorOnStartOfSection < cursorPosition && sectionIndex < state.sections.length) {
-          const section = state.sections[sectionIndex];
-          sectionIndex += 1;
-          cursorOnStartOfSection += `${section.startSeparator}${
-            section.value || section.placeholder
-          }${section.endSeparator}`.length;
-        }
-
-        setSelectedSections(sectionIndex - 1);
-      });
-    } else if (!focused) {
-      setFocused(true);
-      setSelectedSections(sectionOrder.startIndex);
-    } else {
-      const hasClickedOnASection = sectionListRef.current.getRoot().contains(event.target as Node);
-
-      if (!hasClickedOnASection) {
-        setSelectedSections(sectionOrder.startIndex);
+  const getInputContainerClickHandler = useEventCallback(
+    (sectionIndex: number) => (event: React.MouseEvent<HTMLDivElement>) => {
+      // The click event on the clear button would propagate to the input, trigger this handler and result in a wrong section selection.
+      // We avoid this by checking if the call to this function is actually intended, or a side effect.
+      if (event.isDefaultPrevented()) {
+        return;
       }
-    }
-  });
-
-  const handleContainerInput = useEventCallback((event: React.FormEvent<HTMLDivElement>) => {
-    onInput?.(event);
-
-    if (!sectionListRef.current || parsedSelectedSections !== 'all') {
-      return;
-    }
 
-    const target = event.target as HTMLSpanElement;
-    const keyPressed = target.textContent ?? '';
+      setSelectedSections(sectionIndex);
+    },
+  );
 
-    sectionListRef.current.getRoot().innerHTML = state.sections
-      .map(
-        (section) =>
-          `${section.startSeparator}${section.value || section.placeholder}${section.endSeparator}`,
-      )
-      .join('');
-    syncSelectionToDOM();
-
-    if (keyPressed.length === 0 || keyPressed.charCodeAt(0) === 10) {
-      clearValue();
-      setSelectedSections('all');
-    } else if (keyPressed.length > 1) {
-      updateValueFromValueStr(keyPressed);
-    } else {
-      if (parsedSelectedSections === 'all') {
-        setSelectedSections(0);
-      }
-      applyCharacterEditing({
-        keyPressed,
-        sectionIndex: 0,
-      });
-    }
+  const rootProps = useFieldRootProps({
+    manager,
+    internalPropsWithDefaults,
+    stateResponse,
+    applyCharacterEditing,
+    focused,
+    setFocused,
+    domGetters,
   });
 
-  const handleContainerPaste = useEventCallback((event: React.ClipboardEvent<HTMLDivElement>) => {
-    onPaste?.(event);
-    if (readOnly || parsedSelectedSections !== 'all') {
-      event.preventDefault();
-      return;
-    }
+  const handleRootKeyDown = useEventCallback((event: React.KeyboardEvent<HTMLDivElement>) => {
```

Comment:
I'm keeping the `props.onKeyDown?.(event)` (and equivalents) here because on Base UI it's handled automatically so I can't have it in the shared code.